tools/Dashboard/DevHome.Dashboard/ComSafeWidgetObjects/ComSafeWidget.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using DevHome.Common.Extensions;
+using DevHome.Dashboard.Services;
+using Microsoft.UI.Xaml;
+using Microsoft.Windows.Widgets;
+using Microsoft.Windows.Widgets.Hosts;
+using Windows.Foundation;
+
+namespace DevHome.Dashboard.ComSafeWidgetObjects;
+
+/// <summary>
+/// Since Widgets are OOP COM objects, we need to wrap them in a safe way to handle COM exceptions
+/// that arise when the underlying OOP object vanishes. All Widgets should be wrapped in a
+/// ComSafeWidget and calls to the widget should be done through the ComSafeWidget.
+/// This class will handle the COM exceptions and get a new OOP Widget if needed.
+/// All APIs on the IWidget and IWidget2 interfaces are reflected here.
+/// </summary>
+public class ComSafeWidget
+{
+    public Widget OopWidget { get; set; }
+
+    // Not currently used.
+    public DateTimeOffset DataLastUpdated => throw new NotImplementedException();
+
+    public string DefinitionId { get; private set; }
+
+    public string Id { get; private set; }
+
+    // Not currently used.
+    public DateTimeOffset TemplateLastUpdated => throw new NotImplementedException();
+
+    private const int RpcServerUnavailable = unchecked((int)0x800706BA);
+    private const int RpcCallFailed = unchecked((int)0x800706BE);
+
+    public ComSafeWidget(Widget widget)
+    {
+        OopWidget = widget;
+        DefinitionId = widget.DefinitionId;
+        Id = widget.Id;
+        widget.WidgetUpdated += OopWidgetUpdated;
+    }
+
+    public event TypedEventHandler<ComSafeWidget, WidgetUpdatedEventArgs> WidgetUpdated = (_, _) => { };
+
+    private void OopWidgetUpdated(Widget sender, WidgetUpdatedEventArgs args)
+    {
+        WidgetUpdated.Invoke(this, args);
+    }
+
+    public async Task<string> GetCardTemplateAsync()
+    {
+        return await Task.Run(async () =>
+        {
+            try
+            {
+                return await OopWidget.GetCardTemplateAsync();
+            }
+            catch (COMException e) when (e.HResult == RpcServerUnavailable || e.HResult == RpcCallFailed)

```

Comment:
Do we want to at least debug log these failures or are they so common it would be noise?

Reply:
Failures logged.